src/uct/ib/rc/accel/rc_mlx5.inl

```diff
@@ -160,30 +160,92 @@ uct_rc_mlx5_iface_release_srq_seg(uct_rc_mlx5_iface_common_t *iface,
     ++iface->super.rx.srq.available;
 }
 
+#define uct_rc_mlx5_iface_mp_hash_lookup(_h_name, _h_ptr, _key, _last, _flags, \
+                                         _iface) \
+    ({ \
+        uct_rc_mlx5_mp_context_t *ctx; \
+        khiter_t h_it; \
+        int ret; \
+        h_it = kh_get(_h_name, _h_ptr, _key); \
+        if (h_it == kh_end(_h_ptr)) { \
+            /* No data from this sender - this must be the first fragment */ \
+            *(_flags) |= UCT_CB_PARAM_FLAG_FIRST; \
+            if (ucs_likely(_last)) { \
+                /* fast path - single fragment message */ \
+                return &(_iface)->tm.mp.dummy_ctx; \
+            } \
+            h_it = kh_put(_h_name, _h_ptr, _key, &ret); \
+            ucs_assert(ret != 0); \
+            ctx  = &kh_value(_h_ptr, h_it); \
+        } else { \
+            ctx = &kh_value(_h_ptr, h_it); \
+            if (_last) { \
+                (_iface)->tm.mp.dummy_ctx = *ctx; \
```

Comment:
Here we save message context in the dummy, because it is last fragment and we have to remove it.
Using dummy context allows us to remove message context here and do not add any extra branches to remove it later

Reply:
will it work when we receive fragments from multiple endpoints in parallel?

Reply:
yes, because we need to pass this value to UCP (using unexp_eager callback) just once (it is last fragment) and after that we do not need it anymore.

Reply:
so dummy_ctx is actually the context for last fragment?
should it be "last_frag_ctx"?

Reply:
It is used for single-fragment messages as well (which is some sort of last fragment though)
ok, can rename it